File: web/src/main/java/top/bettercode/summer/web/serializer/DivisionSerializer.java
package top.bettercode.summer.web.serializer;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonStreamContext;
import com.fasterxml.jackson.databind.BeanProperty;
import com.fasterxml.jackson.databind.JsonMappingException;
import com.fasterxml.jackson.databind.JsonSerializer;
import com.fasterxml.jackson.databind.SerializerProvider;
import com.fasterxml.jackson.databind.annotation.JacksonStdImpl;
import com.fasterxml.jackson.databind.jsontype.TypeSerializer;
import com.fasterxml.jackson.databind.ser.ContextualSerializer;
import com.fasterxml.jackson.databind.ser.std.StdScalarSerializer;
import java.io.IOException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import top.bettercode.summer.web.serializer.annotation.JsonDivision;
import top.bettercode.summer.web.support.gb2260.Division;
import top.bettercode.summer.web.support.gb2260.GB2260;
import top.bettercode.summer.web.support.gb2260.InvalidCodeException;

/**
 * @author Peter Wu
 */
@JacksonStdImpl
public class DivisionSerializer extends StdScalarSerializer<String> implements
    ContextualSerializer {

  private final Logger log = LoggerFactory.getLogger(DivisionSerializer.class);
  private static final long serialVersionUID = 1L;
  private final boolean vnode;

  public DivisionSerializer() {
    this(false);
  }


  public DivisionSerializer(boolean vnode) {
    super(String.class, false);
    this.vnode = vnode;
  }


  @Override
  public void serialize(String value, JsonGenerator gen, SerializerProvider provider)
      throws IOException {
    gen.writeString(value);

    JsonStreamContext outputContext = gen.getOutputContext();
    String fieldName = outputContext.getCurrentName();
    gen.writeFieldName(fieldName + "Path");
    try {
      Division division = GB2260.getDivision(value);
      gen.writeObject(division.codes(vnode));
    } catch (InvalidCodeException e) {
      log.warn(e.getMessage());
      gen.writeNull();
    }
  }


  @Override
  public JsonSerializer<?> createContextual(SerializerProvider prov, BeanProperty property)
      throws JsonMappingException {
    if (property != null) {
      JsonDivision annotation = property.getAnnotation(JsonDivision.class);
      return new DivisionSerializer(annotation.value());
    }
    return prov.findNullValueSerializer(null);
  }


  @Override
  public boolean isEmpty(SerializerProvider prov, String value) {
    return value.length() == 0;
  }

  @Override
  public final void serializeWithType(String value, JsonGenerator gen, SerializerProvider
      provider,
      TypeSerializer typeSer) throws IOException {
    serialize(value, gen, provider);
  }

}

File: web/src/main/kotlin/top/bettercode/summer/web/serializer/DivisionSerializer.kt
package top.bettercode.summer.web.serializer

import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.databind.BeanProperty
import com.fasterxml.jackson.databind.JsonSerializer
import com.fasterxml.jackson.databind.SerializerProvider
import com.fasterxml.jackson.databind.annotation.JacksonStdImpl
import com.fasterxml.jackson.databind.jsontype.TypeSerializer
import com.fasterxml.jackson.databind.ser.ContextualSerializer
import com.fasterxml.jackson.databind.ser.std.StdScalarSerializer
import org.slf4j.LoggerFactory
import top.bettercode.summer.web.serializer.annotation.JsonDivision
import top.bettercode.summer.web.support.gb2260.GB2260.getDivision
import top.bettercode.summer.web.support.gb2260.InvalidCodeException

/**
 * @author Peter Wu
 */
@JacksonStdImpl
class DivisionSerializer @JvmOverloads constructor(private val vnode: Boolean = false) : StdScalarSerializer<String>(String::class.java, false), ContextualSerializer {
    private val log = LoggerFactory.getLogger(DivisionSerializer::class.java)
    override fun serialize(value: String, gen: JsonGenerator, provider: SerializerProvider) {
        gen.writeString(value)
        val outputContext = gen.outputContext
        val fieldName = outputContext.currentName
        gen.writeFieldName(fieldName + "Path")
        try {
            val division = getDivision(value)
            gen.writeObject(division.codes(vnode))
        } catch (e: InvalidCodeException) {
            log.warn(e.message)
            gen.writeNull()
        }
    }

    override fun createContextual(prov: SerializerProvider, property: BeanProperty?): JsonSerializer<*> {
        if (property != null) {
            val annotation = property.getAnnotation(JsonDivision::class.java)
            return DivisionSerializer(annotation.value)
        }
        return prov.findNullValueSerializer(null)
    }

    override fun isEmpty(prov: SerializerProvider, value: String): Boolean {
        return value.isEmpty()
    }

    override fun serializeWithType(value: String, gen: JsonGenerator, provider: SerializerProvider,
                                   typeSer: TypeSerializer) {
        serialize(value, gen, provider)
    }

    companion object {
        private const val serialVersionUID = 1L
    }
}
